nixos/manual: document kernel backporting policy and implications of it

Ma27 opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done
I'm well aware that this issue is currently under discussion[1] and that these things may change. Also, please don't misinterpret this as an attempt to end the discussion.

This topic made it obvious that people are surprised by the way this issue is handled and only finding out about this unwritten rule because of asking is not a good state IMHO, so I decided to document the following things:

* Right now we drop kernels as soon as they get out of maintenance (LTS kernels even before the next stable NixOS that will exceed their lifespan).
* The `latestCompatibleLinuxPackages` attribute from ZFS isn't monotonic since latest only refers to the latest supported kernel.
* In fact `latestCompatibleLinuxPackages` doesn't seem to be documented at all in the manual, so I also did that.

[1] https://discourse.nixos.org/t/aggressive-kernel-removal-on-eol-in-nixos/23097

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Ma27 committed:
nixos/manual: document kernel backporting policy and implications of it

I'm well aware that this issue is currently under discussion[1] and that
these things may change. Also, please don't misinterpret this as an
attempt to end the discussion.

This topic made it obvious that people are surprised by the way this
issue is handled and only finding out about this unwritten rule because
of asking is not a good state IMHO, so I decided to document the
following things:

* Right now we drop kernels as soon as they get out of maintenance (LTS
  kernels even before the next stable NixOS that will exceed their
  lifespan).
* The `latestCompatibleLinuxPackages` attribute from ZFS isn't
  monotonic since latest only refers to the latest supported kernel.
* In fact `latestCompatibleLinuxPackages` doesn't seem to be documented
  at all in the manual, so I also did that.

[1] https://discourse.nixos.org/t/aggressive-kernel-removal-on-eol-in-nixos/23097

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/aggressive-kernel-removal-on-eol-in-nixos/23097/62

TLATER commented:
Should at least a link to this also be added to the description of `boot.kernelPackages`? I imagine most people who change their kernel versions are more likely to stumble across it there than in the NixOS manual.

NobbZ commented:
As initiator of the linked thread, I am not happy with the status quo, though I appreciate that the status quo will be documented after this got merged and "support" can actually point at the docs, rather than citing from rural knowledge.

Ma27 committed:
nixos/kernel: link manual (with backport policies) from boot.kernelPackages

Yes I know it's dirty to link the rendered HTML, but linking
`#sec-kernel-config` breaks the manual build for me with the following
error:

    element link: validity error : IDREF attribute linkend references an unknown ID "sec-kernel-config"

This is something that should probably be fixed, but for the current
change I'd argue that this is good enough (in fact even the
`nix.settings`-option uses that hack).

Ma27 commented:
@TLATER good idea, done!

(yes I know that linking the rendered manual from the doc is ugly, but apparently we cannot link `#sec-`-sections from the option's doc, cc @pennae for that)

PetarKirov reviewed (commented):
Thanks for taking the initiative to document the status quo. Even if I think that the status quo is suboptimal, I think that it should at least be documented, to help others avoid frustrations caused by using `latestCompatibleLinuxPackages` down the line.

PetarKirov commented on nixos/doc/manual/configuration/linux-kernel.chapter.md:
````diff
@@ -138,3 +148,23 @@ $ cd linux-*
+
+```nix
+{
+  boot.kernelPackages = pkgs.zfs.latestCompatibleLinuxPackages;
````
Thinking about this a bit more...

Given that the manual makes no mention of ZFS, and given the issues that using `latestCompatibleLinuxPackages`, should this really be the first option that we give to users? Imagine a novice user that googles "nixos zfs docs" and this is the first result that pops up (and the only one under `https://nixos.org/manual/nixos/`). Many people would copy this code snippet and not bother reading the warnings that follow. Could we instead suggest something safer, along the lines of:

```suggestion
  
  # Use the current longterm kernel, if ZFS supports it,
  # otherwise use an older one that is the latest supported by ZFS.
  boot.kernelPackages = let
    defaultKernel = pkgs.linuxPackages.kernel
    zfsLatestKernel = pkgs.zfs.latestCompatibleLinuxPackages.kernel
  in
    if zfsLatestKernel.kernelAtLeast defaultKernel.version
    then defaultKernel
    else zfsLatestKernel;
```

SuperSandro2000 replied:
We should not tell people to copy this into their configuration and rather move it to the zfs packages as latestCompatibleLTSLinuxPackages.

ncfavier commented:
Yes, we could add another special prefix in [fix-xrefs.lua](https://github.com/NixOS/nixos-search/blob/main/flake-info/assets/data/fix-xrefs.lua) for `index.html#sec-`.

Ma27 committed:
nixos/doc: suggest using the latest longterm kernel for ZFS

Ma27 commented:
> Given that the manual makes no mention of ZFS, and given the issues that using latestCompatibleLinuxPackages, should this really be the first option that we give to users

That's a good point. Pushed a commit to do that.

> Ideally, this behavior would actually be integrated in latestCompatibleLinuxPackages, while previous meaning could be kept under a compatibility alias and documented that is only meant for use by advanced users.

I think that this makes it even harder to predict which kernel you'll end up with. By default you should always et a longterm kernel (as it's the case currently).

I think that this PR is ready now. If you'd like to propose such a behavior, feel free to open another PR :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/aggressive-kernel-removal-on-eol-in-nixos/23097/71
